pkg/vault/test/foundry/E2eBatchSwap.t.sol

```diff
@@ -0,0 +1,444 @@
+    ERC20TestToken internal tokenC;
+    ERC20TestToken internal tokenD;
+
+    IERC20[] internal tokensToTrack;
+
+    uint256 internal tokenAIdx;
+    uint256 internal tokenBIdx;
+    uint256 internal tokenCIdx;
+    uint256 internal tokenDIdx;
+
+    address internal sender;
+    address internal poolCreator;
+
+    uint256 internal minSwapAmountTokenA;
+    uint256 internal maxSwapAmountTokenA;
+
+    uint256 internal minSwapAmountTokenD;
+    uint256 internal maxSwapAmountTokenD;
+
+    // 0.0001% swap fee
+    uint256 internal constant DEFAULT_SWAP_FEE = 1e12;
+
+    function setUp() public virtual override {
+        BaseVaultTest.setUp();
+
+        IProtocolFeeController feeController = vault.getProtocolFeeController();
+        IAuthentication feeControllerAuth = IAuthentication(address(feeController));
+
+        authorizer.grantRole(
+            feeControllerAuth.getActionId(IProtocolFeeController.setGlobalProtocolSwapFeePercentage.selector),
+            admin
+        );
+
+        _setUpVariables();
+
+        // Set protocol and creator fees to 50%, so we can measure the charged fees.
+        vm.prank(admin);
+        feeController.setGlobalProtocolSwapFeePercentage(FIFTY_PERCENT);
+
+        // Initialize pools that will be used by batch router.
+        // Create poolA
+        vm.startPrank(lp);
+        poolA = _createPool([address(tokenA), address(tokenB)].toMemoryArray(), "poolA");
+        _initPool(poolA, [poolInitAmount, poolInitAmount].toMemoryArray(), 0);
+        // Create poolB
+        poolB = _createPool([address(tokenB), address(tokenC)].toMemoryArray(), "poolB");
+        _initPool(poolB, [poolInitAmount, poolInitAmount].toMemoryArray(), 0);
+        // Create poolC
+        poolC = _createPool([address(tokenC), address(tokenD)].toMemoryArray(), "PoolC");
+        _initPool(poolC, [poolInitAmount, poolInitAmount].toMemoryArray(), 0);
+        vm.stopPrank();
+
+        vm.startPrank(poolCreator);
+        // Set pool creator fee to 100%, so protocol + creator fees equals the total charged fees.
+        feeController.setPoolCreatorSwapFeePercentage(poolA, FixedPoint.ONE);
+        feeController.setPoolCreatorSwapFeePercentage(poolB, FixedPoint.ONE);
+        feeController.setPoolCreatorSwapFeePercentage(poolC, FixedPoint.ONE);
+        vm.stopPrank();
+
+        tokensToTrack = [address(tokenA), address(tokenB), address(tokenC), address(tokenD)].toMemoryArray().asIERC20();
+
+        // Idx of the token in relation to `tokensToTrack`.
+        tokenAIdx = 0;
+        tokenBIdx = 1;
+        tokenCIdx = 2;
+        tokenDIdx = 3;
+    }
+
+    function _createPool(address[] memory tokens, string memory label) internal virtual override returns (address) {
+        address newPool = super._createPool(tokens, label);
+        // Defining swap fees to 0.0001% (Pool Factory Mock sets the swap fee to 0).
+        vault.manualUnsafeSetStaticSwapFeePercentage(newPool, DEFAULT_SWAP_FEE);
```

Comment:
Could we let it be created with 0, then override it in the fuzz tests to also fuzz the swap fee (from 0.0001% to 10%)? Tried it; seems to work.

Reply:
Done!